plugins/providers/virtualbox/action/network_fix_ipv6.rb

```diff
@@ -41,24 +43,38 @@ def call(env)
           # If we have no IPv6, forget it
           return if !has_v6
 
-          # We do, so fix them if we must
-          env[:machine].provider.driver.read_host_only_interfaces.each do |interface|
-            # Ignore interfaces without an IPv6 address
-            next if interface[:ipv6] == ""
+          networks(env).each do |network|
+            next if !present?(network[:ipv6])
+            next if network[:status] != "Up"
 
-            # Make the test IP. This is just the highest value IP
-            ip = IPAddr.new(interface[:ipv6])
-            ip |= IPAddr.new(":#{":FFFF" * (interface[:ipv6_prefix].to_i / 16)}")
+            ip = IPAddr.new(network[:ipv6])
+            ip |= ("1" * (128 - network[:ipv6_prefix].to_i)).to_i(2)
 
             @logger.info("testing IPv6: #{ip}")
+
             begin
               UDPSocket.new(Socket::AF_INET6).connect(ip.to_s, 80)
             rescue Errno::EHOSTUNREACH
               @logger.info("IPv6 host unreachable. Fixing: #{ip}")
-              env[:machine].provider.driver.reconfig_host_only(interface)
+              env[:machine].provider.driver.reconfig_host_only(network)
             end
           end
         end
+
+        # The list of interface names for host-only adapters.
+        # @return [Array<String>]
+        def host_only_interface_names(env)
+          env[:machine].provider.driver.read_network_interfaces
+            .map { |_, i| i[:hostonly] if i[:type] == :hostonly }.compact
+        end
+
+        # The list of networks that are tied to a host-only adapter.
+        # @return [Array]
+        def networks(env)
+          iface_names = self.host_only_interface_names(env)
+          env[:machine].provider.driver.read_host_only_interfaces
+            .select { |network| iface_names.include?(network[:name]) }
+        end
```

Comment:
Nit: I think networks is a bit misleading here - this still returns a list of host_only_interfaces